### PR TITLE
Moved more functionality to *Pattern classes.

### PR DESCRIPTION
**Pull request by knbk**
I've made some additional changes based on your comments on my previous PR. I was actually aiming to get all translation-related code into the `*Pattern` classes, but as it turns out that is difficult to do without affecting the performance of `reverse()` too much.

### CODE REVIEW

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py]: From what I can see, given a specific `*Pattern`, it's *always* called with the same `pattern_name`. As such, it seems to me that the `*Pattern` should be the owner of the `pattern_name` (maybe alongside `URLResolver`/`URLResolver`, maybe instead of).

Some side-remarks about `describe`/`pattern_name`:

* It's only supplied to/for `URLPattern` (and not the `URLResolver`), so only if `is_endpoint=True`.
* The formatting as used in `.describe()` is *almost* the same as used in `technical_404` (the almost being whitespace differences).

Hence: if we make `RoutePattern`/`RegexPattern` owner of `.name`, we can replace `.describe(self)` with `.__str__(self)`, and remove the special casing in `technical_404`.

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py]: Given suggestion above, we should be able to get rid of `pattern_name` as argument here.

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py]: Given suggestion above, we should be able to get rid of `pattern_name` as argument here.

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py]: Given suggestion above, we should be able to get rid of `pattern_name` as argument here.

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py]: *big* assumption here: the `converters` are the same for any translation.

Theoretically, it could be that the converters are different...

```
Dutch: /<decimal_with_comma:amount>/
English: /<decimal_with_period:amount>/
```

(Yes, highly theoretical, but seeing as the converters are part of the translatable string, we should either support choosing having them different, or check for it, not sure).

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py]: Assuming we can change `repr`: `'<%s %s>'.format(self.__class__.__name__, self.pattern.describe(self.name))` would also make sense, and that would become `'<%s %s>'.format(self.__class__.__name__, self.pattern)` under the suggested changes above.

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py:353]: To be honest, I'm not a fan of `default_args` overriding the parsed `kwargs` (and always suspect it's a bug), but let's leave it put because that's how it currently works.

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py]: Would the following be better?

```
sub_path, match_args, match_kwargs = match
```

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py]: Here a minor difference occurs: in the old version, the `default_kwargs` were also run through the converter, in the new version, they're not. I think I prefer the new version, but just mentioning the change.

File: django/contrib/admindocs/views.py
Comment (sjoerdjob) [inline django/contrib/admindocs/views.py]: It's a bit sad that for the new-style patterns, in the admindocs it's processed as "simplified => regex => simplify_regex" losing the (maybe useful?) type information.

(Please don't fix it, it can wait until the base PR is merged, I think).

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py:412]: Seriously wondering why one would even want to support cyclicly nested patterns... From what I can see, the cyclic nature complicates `_populate` (albeit just a bit), but also prevents improvements to the logic of `reverse`. And only using `args` (not kwargs) it brings a possible benefit.

File: django/urls/resolvers.py
Comment (sjoerdjob) [inline django/urls/resolvers.py:373]: Is the problem with cyclic dependencies solved? I *think* so because it has been moved to `pattern` instead.

**Review by sjoerdjob — CHANGES_REQUESTED** [review]
Suggestion: adding `pattern_name` to the `*Pattern` (or `CheckURLMixin`) base class, and the fallout cleanup.

Also: reconsider if we would allow the possibility of locale-dependent converters.

File: django/urls/resolvers.py
Comment (knbk) [reply django/urls/resolvers.py]: I think that rather than localising the route to use different converters, the converters themselves should be localised where necessary. The `regex` attribute on the various converters can be a `LocaleRegexDescriptor` if localisation is needed. If we allow people to use different converters per language, I can only see it end up in a big mess.

If this proves to be a problem, we can add a systems check that iterates `settings.LANGUAGES`. For now I don't think that's necessary.

File: django/urls/resolvers.py
Comment (knbk) [reply django/urls/resolvers.py]: I think semantically, the name belongs to `URLPattern`, but I agree that passing it to the `*Pattern` classes makes for cleaner code. There are quite a lot of places that use `pattern.name`, so I don't think we should remove it from `URLPattern` entirely. 

There's also a different use of quotes around the route, and there are some places that need the route and name separately. I think we need to keep the separation -- quotes around the route in running sentences, no quotes in technical messages with multiple patterns concatenated, and no name when the name is needed separately.

File: django/urls/resolvers.py
Comment (knbk) [reply django/urls/resolvers.py:373]: Yes, it's been solved. The populated attributes on child resolvers are only accessed on non-namespaced includes, while cyclic dependencies are only possible across namespaces.

File: django/urls/resolvers.py
Comment (knbk) [reply django/urls/resolvers.py:412]: I honestly don't think there are any reasonable use-cases, but in my attempts to remove this I encountered cyclic dependencies in Django's own test suite, which raises the barrier to just remove it entirely.

File: django/urls/resolvers.py
Comment (knbk) [reply django/urls/resolvers.py]: Didn't realise this, but I think it makes sense.

File: django/contrib/admindocs/views.py
Comment (knbk) [reply django/contrib/admindocs/views.py]: We can use `base + str(p.pattern)`, so it's an easy fix.

**Review by sjoerdjob — APPROVED** [review]
Thank you for incorporating the changes. Looks good to me.

**Pull request merged**
